### ISSUE TITLE
gccgo: init name collision due to insufficient mangling in Gogo::pkgpath_for_symbol

### ISSUE THREAD

**Issue by thanm** [golang/go#27534]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?

gccgo tip: 
go version devel +dd9e81f678 Thu Aug 30 12:35:33 2018 +0000 linux/amd64

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

linux/amd64

### What did you do?

Unpack the attached tar file (contains a couple of tiny packages and a main) into GOPATH and build with gccgo: 

```
% cd `go env GOPATH`
% tar xvf - < tarfile
initnameclash/
initnameclash/foo/
initnameclash/foo/foo.go
initnameclash_foo/
initnameclash_foo/initnameclash_foo.go
initnameclash_main/
initnameclash_main/main.go
% cd initnameclash_main
% go build -compiler gccgo
```

### What did you expect to see?

successful build

### What did you see instead?

```
% go build -compiler gccgo .
# initnameclash_main
go1: error: duplicate package initialization name ‘initnameclash_foo..import’
go1: note: used by package ‘foo’
go1: note:  and by package ‘initnameclash_foo’
%
```
The clash here is being caused by the current implementation of the compiler's Gogo::pkgpath_for_symbol() function, which (in the process of trying to generated an assembler-compatible symbol) can map two distinct package paths to the same symbol name; this triggers a name collision later on.

Tar file with sources here:
[tarfile.gz](https://github.com/golang/go/files/2357664/tarfile.gz)

**Comment by ianlancetaylor** [comment]
I just want to note that if we change `Gogo::pkgpath_for_symbol`, we have to change at least `(*Package).gccgoSymbolPrefix` in cmd/cgo/out.go.  I expect the same algorithm appears in other places as well.

**Comment by thanm** [comment]
Thanks for the pointer. 

Other places: I agree, any change to the algorithm in `Gogo::pkgpath_for_symbol` could result in large numbers of knock-on changes (especially in go:linkname directives in Go sources and __asm__ directives in C sources). There are many such spots that would have to be edited.

Here is a modest proposal: let's say we change the semantics of the "//go:linkname" directive for gccgo. Right now it operates in a purely mechanical fashion, e.g. when you write

`  //go:linkname ARG1 ARG2
`

the compiler forces symbol ARG1 to be exported as exactly ARG2. How about if we change the rules as follows:

 - if the second argument of //go:linkname (ARG2) does NOT contain a '.', do as before (ARG1 is renamed and exported as ARG2)

 - if ARG2 does contain a '.', interpret ARG2 as a package-qualified name, and export ARG1 to the symbol formed by running Gogo::pkgpath_for_symbol on that package qualified name.

In other words, if I write:

` //go:linkname my_c_function_util xyz/pack.function_util
`

then the compiler will take "xyz/pack.function_util" and treat it as "`<packagepath>.<symbol>`" and run it through `Gogo::pkgpath_for_symbol` to arrive at the correct assembler symbol.

An advantage of doing things this way is that it could reduce the number of differences between the libgo version and the main GC version of a package. For example in "runtime/atomic_pointer.go" right now:

GCCGO:   //go:linkname sync_atomic_StoreUintptr sync_atomic.StoreUintptr
GC:      //go:linkname sync_atomic_StoreUintptr sync/atomic.StoreUintptr

With this change we could just use the GC style directive.

**Comment by ianlancetaylor** [comment]
I'm a little bit reluctant to pass anything other than a pkgpath to `Gogo::pkgpath_for_symbol`.  But I don't know of any way to reliably pick out the package name from a symbol name.

I'm also a little bit reluctant to do any operation at all on a `//go:linkname` directive, since the current behavior is easy to understand and any change could be rather cryptic.

**Comment by thanm** [comment]
I've sent a CL with a fix.

I am a little nervous about this change given that it will make debugging more difficult (to print a global variable you will have to know its mangled name), but it does solve the collision problem.
